torch/distributed/_sharded_tensor/api.py

```diff
@@ -750,3 +796,61 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
                           device=local_device, )
     else:
         raise ValueError(f'Unsupported create_op: {tensor_init_params.create_op}')
+
+
+class PartialTensor(object):
+    """
+    PartialTensor is an abstraction to represent Tensors that need
+    aggregation across multiple devices and multiple processes.
+
+    PartialTensor is initialized in an SPMD like fashion where each rank
+    initializes the PartialTensor. The PartialTensor object on each rank
+    then only stores the local partial shard and provides global metadata
+    for all the shards.
+
+    PartialTensor doesn't provide any Tensor like operations but is a
+    wrapper providing the Tensor representing the local partial shard
+    and the global metadata.
+
+    Using these, users can apply custom distributed sharded computations
+    on top of this primitive. The local shards are all initialized using
+    the create_op specified by tensor_init_params.create_op, e.g.,
+    torch.ones, or torch.empty.
+
+    Args:
+        local_partial_results (List(Tensor)):
+            Partial results are stored across ranks and in a form of a list.
+    """
+
+    def __init__(self, local_partial_results):
+        self.local_partial_results = local_partial_results
+
+    def __post_init__(self):
+        if not isinstance(self.local_partial_results, Iterable):
+            raise ValueError('local_partial_results needs to be a Iterable.')
+        for tensor in self.local_partial_results:
+            if not isinstance(tensor, torch.Tensor):
+                raise ValueError('Element in local_partial_results needs to be a Tensor.')
+
+    def reshard(self, _resharding_spec: ReshardingSpec) -> torch.Tensor:
```

Comment:
nit: why does `_resharding_spec` have an underscore? (usually this is done for hidden params)

Reply:
Oh, with the enum added, we will be using it.